src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt

```diff
@@ -39,8 +39,11 @@ object RsCommonCompletionProvider : CompletionProvider<CompletionParameters>() {
         context: ProcessingContext,
         result: CompletionResultSet
     ) {
-        val element = parameters.position.parent as RsReferenceElement
-        if (parameters.position !== element.referenceNameElement) return
+        // Use original position if possible to re-use caches of the real file
+        val position = CompletionUtil.getOriginalOrSelf(parameters.position)
```

Comment:
Looks like we should accurately handle cases when `parameters.position` and `position` have different token types to avoid exceptions described [here](https://github.com/intellij-rust/intellij-rust/pull/3359#issuecomment-461341968) and still provide correct completion
I think it's worth to add test(s) for such case to commit right behavior

Reply:
Positions can't have different types b/c they are always leaf elements (and the same type is guaranteed by `CompletionUtil.getOriginalOrSelf`), BUT they can be children of different types. And [here](https://github.com/intellij-rust/intellij-rust/pull/3359#issuecomment-461341968) we failed on the parent element cast. Maybe we should rather check that all ancestors of both elements have the same type?

Reply:
> Positions can't have different types b/c they are always leaf elements

Yes, but they can have different **token types**, for example, `identifier` and `let` keyword => psi structures of original file and completion copy one differ

> Maybe we should rather check that all ancestors of both elements have the same type?

I think it's enough to just check that parents have the same type or that `CompletionUtil#getOriginalOrSelf` and `parameters.position` have the same token type (looks like these conditions always have the same value)